lighthouse-core/audits/unsized-images.js

```diff
@@ -0,0 +1,127 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./audit.js');
+const i18n = require('./../lib/i18n/i18n.js');
+const URL = require('./../lib/url-shim.js');
+
+const UIStrings = {
+  /** Title of a Lighthouse audit that provides detail on whether all images have explicit width and height. This descriptive title is shown to users when every image has explicit width and height */
+  title: 'Image elements have explicit `width` and `height`',
+  /** Title of a Lighthouse audit that provides detail on whether all images have explicit width and height. This descriptive title is shown to users when one or more images does not have explicit width and height */
+  failureTitle: 'Image elements do not have explicit `width` and `height`',
+  /** Description of a Lighthouse audit that tells the user why they should include explicit width and height for all images. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. */
+  description: 'Always include explicit width and height on your image elements to reduce layout shifting and improve CLS. [Learn more](https://web.dev/optimize-cls/#images-without-dimensions)',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/**
+ * @fileoverview
+ * Audit that checks whether all images have explicit width and height.
+ */
+
+class SizedImages extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'unsized-images',
+      title: str_(UIStrings.title),
+      failureTitle: str_(UIStrings.failureTitle),
+      description: str_(UIStrings.description),
+      requiredArtifacts: ['ImageElements'],
+    };
+  }
+
+  /**
+   * @param {string} attr
+   * @return {boolean}
+   */
+  static isValidAttr(attr) {
+    // An img size attribute is valid for preventing CLS
+    // if it is a non-negative, non-zero integer.
+    const NON_NEGATIVE_INT_REGEX = /^\d+$/;
+    const ZERO_REGEX = /^0+$/;
+    return NON_NEGATIVE_INT_REGEX.test(attr) && !ZERO_REGEX.test(attr);
+  }
+
+  /**
+   * @param {string} property
+   * @return {boolean}
+   */
+  static isValidCss(property) {
+    // An img css size property is valid for preventing CLS
+    // if it is defined, not empty, and not equal to 'auto'.
+    // `undefined` and `''` are implicitly rejected as invalid
```

Comment:
nit: undefined is never given to this function (not reflected in type either) because the calling code excludes falsy values from calling this function.

maybe remove that falsy check in the calling code? or just tweak this comment to drop mentioning undefined? could just explicitly mention "falsy values are rejected as invalid..."

Reply:
thanks, the cssHeight/cssWidth collected in the gatherer can be `string | undefined`, but I forgot to update the type for this function because of the (lack of) calling behavior you mentioned

However, I wanted to keep isValidCss / isValidAttr to be standalone functions so I think I'll 
1. remove the falsy check in isUnsizedImage for both isValidAttr/isValidCss 
2. tweak the comment
3. update the type for @param in isValidCss